### DISCUSSION TITLE
Possible issue with handling Abort in bacpoll / bac-discover / bac-rw

### DISCUSSION THREAD

**Discussion by nazaryev-cool** [bacnet-stack/bacnet-stack#726] [General]
tl;dr: It seems that the bac-discover component doesn't handle cases where the device responds to ReadPropertyMultiple request with an abort correctly. Instead of generating an error or falling back to a ReadProperty request, the device's Discovery_State remains stuck in BACNET_DISCOVER_STATE_OBJECT_GET_PROPERTY_REQUEST indefinitely.

What exactly happens in my case:
1. [bac-discover sends a ReadPropertyMultiple request via bac-rw.](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/client/bac-discover.c#L805)
2. The device responds with an Abort message indicating "segmentation-not-supported."
3. [bac-rw's MyAbortHandler is called, and global variable Error_Detected sets to true](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/client/bac-rw.c#L118)
4. [The invoke ID associated with the ReadPropertyMultiple request is freed.](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/service/h_apdu.c#L727)
5. [bac-rw clears the Error_Detected global variable.](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/client/bac-rw.c#L469)
6. [bac-rw concludes there is no error and decides not to call bacnet_read_write_value_callback.](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/client/bac-rw.c#L526)
7. bac-discover fails to recognize that no response was received, and the device remains in the BACNET_DISCOVER_STATE_OBJECT_GET_PROPERTY_REQUEST state indefinitely.

My workaround for this situation is just to move [this if-clause](https://github.com/bacnet-stack/bacnet-stack/blob/dab6d8af5506c8163c8942808ffa3ffede0c89c3/src/bacnet/basic/client/bac-rw.c#L477) up, before `tsm_invoke_id_free(Request_Invoke_ID)` and in such case bac-discover (and any other bac-rw users) will receive the error.
Anyway, it's still unclear what to do with this error in bac-discover or bacpoll, and, if bac-discover is just an sample and not a "production-ready" component, it's not too important to somehow fix it.

Disclaimer: I'm not 100% sure that I properly diagnosed the issue (that's why I write it here waiting for some feedback) and, of course, I don't know if my workaround doesn't break anything else. Plus I just started to tinker with bacnet-stack, so please don't be too harsh :)

**Comment by skarg** [comment]
Thank you for reporting the problem and suggesting a partial solution.

I wrote bac-rw.c as a production ready re-usable module, but it doesn't include any fallback to ReadProperty when ReadPropertyMultiple is not supported by the downstream device.  For my use case, all the downstream devices supported ReadPropertyMultiple so I didn't see the problem.

The user code - in this case bac-discover.c module - would need to respond to the error and request either the Property_List property (similar to the Device Object_List property) or use a static list of properties to query from that object. The apps/epics/main.c module does this, but I haven't made the time to add the extra handling into bac-discover.c module yet.

**Reply by skarg** [reply]
Fixed endless query in bac-rw module when error is returned. [#727]

**Reply by nazaryev-cool** [reply]
Thank you for your work!

> request either the Property_List property

From what I understand, Property_List is a relatively new feature (introduced sometime between 2008 and 2016). It turns out that the third-party device I'm working with (and I expect that it would be much more than this one) doesn't have this feature. So, the code should ideally use RPM with PROP_ALL first. If that's not available, it should fall back to Property_List, and if that's not available either, it should only use the required properties (property_list_required).

> but I haven't made the time to add the extra handling into bac-discover.c module yet.

If you are not planning to implement it in the nearest future, are you interested to let me try to implement it and submit a pull request? I'm not sure I'll succeed or my implementation will meet your high level coding standards, but I'd like to try.

**Reply by skarg** [reply]
Yes, property-list was added to BACnet in 2012, although I proposed it in 2007 (the XML working groups didn't see the usefulness until a few years later). I also proposed a writable-property-list property a few years ago and it is still pending committee approval.  Some things - especially small things - take a long time to standardize.

You are welcome to implement the 'property-list' feature - probably easiest for the moment is the fallback to a static list of required properties, and the library already includes one in property.c module.  I had already set up the data structure to hold each object property list in bac-discover.c so there should be some code already started:
```
typedef struct bacnet_object_data_t {
    OS_Keylist Property_List;
    /* used for discovering object data */
    uint32_t Property_List_Size;
    uint32_t Property_List_Index;
} BACNET_OBJECT_DATA;
```

**Comment by nazaryev-cool** [comment]
I've found one more very similar issue in bac-rw. Should I create a dedicated discussion thread or even a GitHub issue next time, btw?

When trying to read a property from a non-existent object, we might get a property-access-error in response. However, the bac-rw code ignores this error and doesn’t trigger the callback with the error.

Steps to reproduce:

1. Send a ReadPropertyMultiple request to device 1, specifying object device:999 and PROP_ALL.
2. The device responds with a `property-access-error`, with `error-class: object (1)` and `error-code: unknown object (31)`.
3. To handle the response, the [My_Read_Property_Multiple_Ack_Handler][rpm-handler] in bac-rw is called, which then calls rpm_ack_object_property_process, passing bacnet_read_property_ack_process as the callback.
4. Inside rpm_ack_object_property_process, we enter the [error handling section][property-error] where rp_data->error-class = 1 and rp_data->error-code = 31 are set, and then bacnet_read_property_ack_process is called.
5. At this [point](https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/basic/client/bac-rw.c#L212), there is no check to see if errors have already been recorded in rp_data. The apdu_len is 0, so we don’t enter the while(apdu_len) loop, meaning the callback never gets triggered.

The easiest way to fix it is to add something like this before the `while(apdu_len)` loop:
```
if (rp_data->error_code != ERROR_CODE_SUCCESS) {
    if (bacnet_read_write_value_callback) {
         bacnet_read_write_value_callback(device_id, rp_data, NULL);
    }
}
```
... and also initialize all `rp_data` ([1][rpdata-1], [2][rpdata-2], [3][rpdata-3]) to be sure that error_code is set to something reasonable (e.g. ERROR_CODE_SUCCESS (84?)). Otherwise, there is a big chance that rp_data->error_code could contain garbage from the stack (that's what happened to me, btw).

[rpm-handler]: https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/basic/client/bac-rw.c#L311
[property-error]: https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/rpm.c#L717

[rpdata-1]: https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/basic/client/bac-rw.c#L272
[rpdata-2]: https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/basic/client/bac-rw.c#L305
[rpdata-3]: https://github.com/bacnet-stack/bacnet-stack/blob/0177c59f4ac9e642b45f25991da7168b4b7c9f0d/src/bacnet/basic/client/bac-rw.c#L520

**Reply by skarg** [reply]
Thank you for reporting the problem, steps to reproduce, and some suggested fixes! I put those fixes into a pull request [#756] - maybe you can review those changes to see if they fix the problem for you?

**Reply by nazaryev-cool** [reply]
Thanks for getting back to me so quickly! I can confirm that the previous issue is resolved. However, I've run into another similar problem, and I'm not sure how to fix it.

Here's what I've found:

1. I send a ReadProperty request with PROP_DEVICE_ADDRESS_BINDING to the device.
2. The device responds with an empty list (3E 3F).
3. In the datalink_receive() function, it goes through npdu_handler(pdu_len=14), then apdu_handler(apdu_len=12), and finally My_Read_Property_Ack_Handler(service_len=9). At this point, we call rp_ack_decode_service_request(service_len/apdu_size=9) to fill rp_data.
4. Inside rp_ack_decode_service_request(), the arguments apdu and apdu_size are used, which is confusing because we're dealing with a part of the APDU (object-identifier + property-identifier + property-value), and its size is 3 less than the actual apdu_len. Anyway, we [calculate the property-value size][calculate-size], get 0, set application_data_len to 0, and return to My_Read_Property_Ack_Handler.
5. My_Read_Property_Ack_Handler then calls bacnet_read_property_ack_process(rp_data).
6. bacnet_read_property_ack_process creates two local variables, apdu and apdu_len, and assigns them to application_data and application_data_len respectively. Now, apdu points to the property-value or more specifically, to the closing tag (0x3F), and apdu_len is 0 (this is still confusing because apdu_len has had 3 different meanings in the same code path). And as apdu_len equals 0, we can't enter the `while(apdu_len)` loop, and even if we could, bacapp_decode_known_property → [bacnet_is_closing_tag][is-closing-tag] will return false as `apdu_len` is equal 0.

I’m sorry for writing so often. I hope it’s not too much of a bother.

[calculate-size]: https://github.com/bacnet-stack/bacnet-stack/blob/f806c5829be1c40e014a6ef51b1fd0f0ed506258/src/bacnet/rp.c#L475
[is-closing-tag]: https://github.com/bacnet-stack/bacnet-stack/blob/f806c5829be1c40e014a6ef51b1fd0f0ed506258/src/bacnet/bacdcode.c#L637

**Reply by nazaryev-cool** [reply]
P.S. I also noticed that this problem doesn’t happen with ReadPropertyMultiple. However, after a closer look, it turns out this is just a coincidence. It also has a [calculation for the property-value size][rpm-calc-size], and for an empty-list it assigns application_data_len to 0, and because of that `rp_data->application_data` and `rp_data->application_data_len` are not [overridden here][no-override] and contain the data from _previous_ parsed property, even though rp_data->object_property is [changed here][rpm-parse-prop] to PROP_DEVICE_ADDRESS_BINDING. So, in the end the code calls callback with `rp_data` that points to the previous property value and current property id :)

[rpm-calc-size]: https://github.com/bacnet-stack/bacnet-stack/blob/f806c5829be1c40e014a6ef51b1fd0f0ed506258/src/bacnet/rpm.c#L697
[no-override]: https://github.com/bacnet-stack/bacnet-stack/blob/f806c5829be1c40e014a6ef51b1fd0f0ed506258/src/bacnet/rpm.c#L703
[rpm-parse-prop]: https://github.com/bacnet-stack/bacnet-stack/blob/f806c5829be1c40e014a6ef51b1fd0f0ed506258/src/bacnet/rpm.c#L687

**Reply by nazaryev-cool** [reply]
@skarg when you have a moment, could you please provide some feedback on this? I’d really appreciate it!

**Reply by skarg** [reply]
Thank you for tracing the to find the problems.  The EMPTY_LIST tag is a newer edition, but can still be useful for the empty lists. I also noticed a number of places where error handling should induce a callback. I put some fixes in a pull-request [#765].
